### PR TITLE
Add Sphinx documentation generation functionality

### PR DESCRIPTION
**Pull request by gsingh93**
This PR adds functionality to generate local HTML docs:
image

Most of these files were generated with `sphinx-quickstart`, after which I made some minor modifications. To use it, install the few requirements from `requirements.txt`. Then run `make html`. Open `build/html/index.html` in a web browser to browse the generated documentation.

There are a few warnings that are currently shown that result in some modules not being generated:
```
WARNING: autodoc: failed to import module 'browser'; the following exception was raised:
attempted relative import beyond top-level package
WARNING: autodoc: failed to import module 'anki2' from module 'importing'; the following exception was raised:
No module named 'importing.anki2'; 'importing' is not a package
WARNING: autodoc: failed to import module 'apkg' from module 'importing'; the following exception was raised:
No module named 'importing.apkg'; 'importing' is not a package
WARNING: autodoc: failed to import module 'base' from module 'importing'; the following exception was raised:
No module named 'importing.base'; 'importing' is not a package
WARNING: autodoc: failed to import module 'csvfile' from module 'importing'; the following exception was raised:
No module named 'importing.csvfile'; 'importing' is not a package
WARNING: autodoc: failed to import module 'mnemo' from module 'importing'; the following exception was raised:
No module named 'importing.mnemo'; 'importing' is not a package
WARNING: autodoc: failed to import module 'noteimp' from module 'importing'; the following exception was raised:
No module named 'importing.noteimp'; 'importing' is not a package
WARNING: autodoc: failed to import module 'pauker' from module 'importing'; the following exception was raised:
No module named 'importing.pauker'; 'importing' is not a package
WARNING: autodoc: failed to import module 'supermemo_xml' from module 'importing'; the following exception was raised:
No module named 'importing.supermemo_xml'; 'importing' is not a package
```

But the majority of the modules work fine, and there's nothing I hate more than debugging Python import issues, so I'm hoping we can get this PR merged and figure out this particular issue later.

### CODE REVIEW

File: sphinx-docs/source/conf.py [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+#
+# For the full list of built-in configuration values, see the documentation:
+# https://www.sphinx-doc.org/en/master/usage/configuration.html
+
+import os
+import sys
+
+project = 'Anki'
+copyright = '2023, dae'
+author = 'dae'
```
Comment (gsingh93) [inline sphinx-docs/source/conf.py:11]: Is this correct or should it be changed?

File: python/sphinx/conf.py [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+# Configuration file for the Sphinx documentation builder.
+#
+# For the full list of built-in configuration values, see the documentation:
+# https://www.sphinx-doc.org/en/master/usage/configuration.html
+
+import os
+import sys
+
+project = 'Anki'
+copyright = '2023, dae'
```
Comment (dae) [inline python/sphinx/conf.py:12]: Please use 'Ankitects Pty Ltd and contributors'

**Review by dae — COMMENTED** [review]
I've converted the makefiles to use Anki's build system. You can invoke it with

./ninja python:sphinx

optionally adding -v to see the command that's run

Add the following if you want to see all output

```
    fn hide_success(&self) -> bool {
        false
    }
```

It doesn't seem to be working for me - the paths may need updating?

**Comment by gsingh93** [comment]
Thanks for handling the ninja part of this, it would have taken me a while to figure that out.

> It doesn't seem to be working for me

There was one step I originally did manually that I forgot to include in the original commit: `sphinx-apidoc` needs to be run before `sphinx-build`. I've updated the build file to do this, and now if you open `out/python/sphinx/index.html`, things seem to be working.

There's one open question left here: `sphinx-apidoc` puts a bunch of files in `python/sphinx`, i.e. `python/sphinx/collection.rst`:
```
collection module
=================

.. automodule:: collection
   :members:
   :undoc-members:
   :show-inheritance:
```

These show up as uncommitted files after running a build. This is my first time using Sphinx so I'm not very familiar with it, but my understanding is that normally you would run `sphinx-apidoc` to generate these files, and then you can customize these `.rst` files with more information other than just the auto-generated module documentation. In this case, you would commit these to the repo. My main goal with this PR was just to get the module documentation, so I don't know if anyone will be interested in modifying these files in the future.

So we could commit these to the repo, or we could add them to .gitignore, or we could copy `python/sphinx` to some `out/python/sphinx/staging` directory (is there a built-in `ninja` way to do this or would we just call `cp`?), make `apidoc` generate the docs in that staging directory, and then run `sphinx-build` with the staging directory as the source directory (instead of `python/sphinx`). I don't have any preference here, let me know what you think makes the most sense.

**Comment by dae** [comment]
I've reworked things so that the build files are all written to out/, and used autoapi so the correct anki/aqt prefix is added to the modules. If you're happy with it, I'll merge it in.

File: python/sphinx/conf.py [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+project = 'Anki'
+copyright = '2023, Ankitects Pty Ltd and contributors'
+author = 'Ankitects Pty Ltd and contributors'
+
+REPO_ROOT = os.environ["REPO_ROOT"]
+
+extensions = ['sphinx_rtd_theme', 'autoapi.extension']
+html_theme = 'sphinx_rtd_theme'
+autoapi_python_use_implicit_namespaces = True
+autoapi_dirs = [os.path.join(REPO_ROOT, x) for x in ["pylib/anki", "out/pylib/anki", "qt/aqt", "out/qt/aqt"]]
```
Comment (gsingh93) [inline python/sphinx/conf.py:19]: I'm getting this error:
```
Extension error:
AutoAPI Directory `/Users/gsgx/code/anki/out/qt/aqt` not found. Please check your `autoapi_dirs` setting.
Traceback (most recent call last):
  File "/Users/gsgx/code/anki/python/sphinx/build.py", line 7, in <module>
    subprocess.run(["out/pyenv/bin/sphinx-build", "out/python/sphinx", "out/python/sphinx/html"], check=True)
  File "/Users/gsgx/code/anki/out/extracted/python/lib/python3.9/subprocess.py", line 528, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '['out/pyenv/bin/sphinx-build', 'out/python/sphinx', 'out/python/sphinx/html']' returned non-zero exit status 2.
Failed with code Some(1): /Users/gsgx/code/anki/out/pyenv/bin/python python/sphinx/build.py
```

`out/qt/aqt` doesn't exist for me but `out/qt/_aqt` does, is this what you meant to add here?

Comment (dae) [reply python/sphinx/conf.py:19]: Let's just remove it - by the looks of it, it's empty apart from a tsconfig file.

Comment (gsingh93) [reply python/sphinx/conf.py:19]: I do see some code in that directory:
```
out/qt/_aqt/forms:
__pycache__      addmodel_qt6.py     browserdisp_qt6.py  clayout_top_qt6.py  editaddon_qt6.py    exporting_qt6.py      findreplace_qt6.py  main_qt6.py         preview_qt6.py     setgroup_qt6.py   synclog_qt6.py
about_qt5.py     addonconf_qt5.py    browseropts_qt5.py  customstudy_qt5.py  editcurrent_qt5.py  fields_qt5.py         forget_qt5.py       modelopts_qt5.py    profiles_qt5.py    setlang_qt5.py    taglimit_qt5.py
about_qt6.py     addonconf_qt6.py    browseropts_qt6.py  customstudy_qt6.py  editcurrent_qt6.py  fields_qt6.py         forget_qt6.py       modelopts_qt6.py    profiles_qt6.py    setlang_qt6.py    taglimit_qt6.py
addcards_qt5.py  addons_qt5.py       changemap_qt5.py    dconf_qt5.py        edithtml_qt5.py     filtered_deck_qt5.py  getaddons_qt5.py    models_qt5.py       progress_qt5.py    stats_qt5.py      template_qt5.py
addcards_qt6.py  addons_qt6.py       changemap_qt6.py    dconf_qt6.py        edithtml_qt6.py     filtered_deck_qt6.py  getaddons_qt6.py    models_qt6.py       progress_qt6.py    stats_qt6.py      template_qt6.py
addfield_qt5.py  browser_qt5.py      changemodel_qt5.py  debug_qt5.py        emptycards_qt5.py   finddupes_qt5.py      importing_qt5.py    preferences_qt5.py  reposition_qt5.py  studydeck_qt5.py  widgets_qt5.py
addfield_qt6.py  browser_qt6.py      changemodel_qt6.py  debug_qt6.py        emptycards_qt6.py   finddupes_qt6.py      importing_qt6.py    preferences_qt6.py  reposition_qt6.py  studydeck_qt6.py  widgets_qt6.py
addmodel_qt5.py  browserdisp_qt5.py  clayout_top_qt5.py  editaddon_qt5.py    exporting_qt5.py    findreplace_qt5.py    main_qt5.py         preview_qt5.py      setgroup_qt5.py    synclog_qt5.py
```

But I'm assuming this isn't relevant for developers, in which case I agree we can just remove this directory

Comment (dae) [reply python/sphinx/conf.py]: The error you're getting is for out/qt/aqt, not out/qt/_aqt. I've updated the path in conf.py - does it work for you now?

**Comment by gsingh93** [comment]
I tested the latest commit and everything seems to be working, so this should be good to merge. Thanks!

Also, where should this be documented, https://github.com/ankitects/anki/blob/main/docs/development.md?

**Pull request merged**